src/app/CommandSender.h

```diff
@@ -181,29 +181,98 @@ class CommandSender final : public Messaging::ExchangeDelegate
+    // to make necessary changes for new functionality in a single location.
+    struct AddRequestDataParameters
     {
         // gcc bug requires us to have the constructor below
         // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96645
-        AdditionalCommandParameters() {}
+        AddRequestDataParameters() {}
 
-        AdditionalCommandParameters & SetStartOrEndDataStruct(bool aStartOrEndDataStruct)
+        AddRequestDataParameters(const Optional<uint16_t> & aTimedInvokeTimeoutMs) : timedInvokeTimeoutMs(aTimedInvokeTimeoutMs) {}
+
+        // When a value is provided for timedInvokeTimeoutMs, this invoke becomes a timed
+        // invoke. CommandSender will use the minimum of all provided timeouts for execution.
+        const Optional<uint16_t> timedInvokeTimeoutMs;
+        // The command reference is required when sending multiple commands. It allows the caller
+        // to associate this request with its corresponding response.
+        Optional<uint16_t> commandRef;
+    };
+
+    // PrepareCommand is a public SDK API, so we must maintain source compatibility.
+    // Using this struct for API parameters instead of individual parameters allows us
+    // to make necessary changes for new functionality in a single location.
+    struct PrepareCommandParameters
+    {
+        // gcc bug requires us to have the constructor below
+        // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96645
+        PrepareCommandParameters() {}
+
+        PrepareCommandParameters(const AddRequestDataParameters & aAddRequestDataParam) :
+            commandRef(aAddRequestDataParam.commandRef)
+        {}
+
+        PrepareCommandParameters & SetStartDataStruct(bool aStartDataStruct)
         {
-            startOrEndDataStruct = aStartOrEndDataStruct;
+            startDataStruct = aStartDataStruct;
             return *this;
         }
 
-        // From the perspective of `PrepareCommand`, this is an out parameter. This value will be
-        // set by `PrepareCommand` and is expected to be unchanged by caller until it is provided
-        // to `FinishCommand`.
+        PrepareCommandParameters & SetCommandRef(uint16_t aCommandRef)
+        {
+            commandRef.SetValue(aCommandRef);
+            return *this;
+        }
+        // The command reference is required when sending multiple commands. It allows the caller
+        // to associate this request with its corresponding response. We validate the reference
+        // early in PrepareCommand, even though it's not used until FinishCommand. This proactive
+        // validation helps prevent unnecessary writing an InvokeRequest into the packet that later
+        // needs to be undone.
+        //
+        // Currently, provided commandRefs for the first request must start at 0 and increment by one
+        // for each subsequent request. This requirement can be relaxed in the future if a compelling
+        // need arises.
```

Comment:
Here's the thing: the reason we generated these inside CommandSender (after the initial attempts at this which did pass them in) is so that we can't mess it up and consumers who don't care about the refs don't have to worry about generating them.

Why do we now want to (1) force consumers to generate the refs but (2) force them to do it in a particular way which we could just as easily do?

Reply:
But most importantly: we are changing something that was explicitly a review comment on an earlier PR to the thing that review comment asked us not to do...  Maybe there is new information now, but it's sort of odd to do that.

Reply:
>consumers who don't care about the refs don't have to worry about generating them.

These consumers need to care at some level, otherwise they are not going to be able to associate the response with the request they sent out. If they really don't care they won't be using batch commands. Right?

1. This way, the caller knows the reference associated with the request prior to adding the request, instead of waiting for request to be added successfully and then reading the value from an out param. This should make it easier for higher level tracking responses to requests without extra back channel communication over various JNI/CFFI type interfaces what the CommandRef was.
2. As mentioned in the comment here, this requirement can be relaxed in the future if we build up a tracking mechanism in CommandSender that is behind the `CHIP_CONFIG_SENDING_BATCH_COMMANDS_ENABLED`. We are using a primitive way of tracking because of the low overhead, but we have the flexibility of loosening this should there be a compelling case without breaking this current scheme for any client that implements it this way right now.

You are referring to https://github.com/project-chip/connectedhomeip/pull/30675#discussion_r1414883100, right?
I had a chat with possible consumers of this API and they didn't like the notion of having to pass this reference back up the call chain.

It seems like the main reason you'd like to see CommandSender doing this is because:
>If we leave it to API consumers, it leaves more room for creating invalid requests

Why is this such a concern? We have a mechanism in place right now that, although currently is restrictive could be relaxed, that prevents use from sending out an InvokeRequestMessage that contains duplicate CommandRefs.

Outside of "leaves more room for creating invalid requests", is there any other concern about having consumer provide the CommandRef?

Reply:
> These consumers need to care at some level,

They care what the refs are.  They don't care to write a ref allocation policy that follows whatever rules the spec has....

> This should make it easier for higher level tracking responses to requests without extra back channel communication over various JNI/CFFI type interfaces what the CommandRef was.

I guess it depends on what the upper layer APIs look like, but what I was expecting is that there are basically two types of high-level APIs here.

1. Explicit batch: the API takes a list of commands, and sends them as batch.  It returns, async, a list of responses in the same order.  That means it needs to insert the responses into the right place in the list, so it needs to keep track of what the order of the relevant refs was, right?
2. Implicit batch: the API takes commands, each command with its own "call me async" thing, and then batches them under the hood.  In this case it needs to track which ref corresponds to which "call me async" thing.

I guess I am not seeing what the "back channel" bits here are and why they are a problem, particularly....

> this requirement can be relaxed in the future

But why are we forcing all clients to do it a particular way right now?  If we want to force this particular way, we should just tell the clients what the refs are.  If we want clients to be able to pick any refs themselves, fine, but we are not doing that....

> I had a chat with possible consumers of this API and they didn't like the notion of having to pass this reference back up the call chain.

I'm a (not possible, very likely) consumer of this API and I don't like the notion of having to write logic to generate these refs and then maintain it.

Reply:
>I guess I am not seeing what the "back channel" bits here are and why they are a problem, particularly....

You either need to backchannel the `CommandRef` or introducing tracking in the glue code that goes whatever language you are in to the c++ implementation

Lets use chip-repl as an example. Right now [command.cpp](https://github.com/project-chip/connectedhomeip/blob/f66ce35c9e876469612a27af89e45248bf731b4c/src/controller/python/chip/clusters/command.cpp#L120) (side note, yes I am aware I need to fix `GetIndexFrocommandRef` -> `GetIndexFromCommandRef`) needs to keep track of the commandRef to index. After the translation to an index this then is forwarded to [Command.py](https://github.com/project-chip/connectedhomeip/blob/f66ce35c9e876469612a27af89e45248bf731b4c/src/controller/python/chip/clusters/Command.py#L148). The python side then hold onto the response in an array until the OnDone callback where it give the response in order.

If python side could dictate the CommandRef, we could eliminate the need for tracking in the interim to do CommandRef to array index value. While in this case we don't propagate the CommandRef back to python, some implementations might want a thin glue layer between SDK APIs and would want to do the tracking themselves thus needing a way to "back channel" the CommandRef.

Additionally for cert testing automation we actually need to be able to dictate the commandRef. We want to intentionally send duplicate CommandRefs to make sure DUT fails. Note this should always be behind a TestOnly function and I was explicitly adding that in this PR.

>I'm a (not possible, very likely) consumer of this API and I don't like the notion of having to write logic to generate these refs and then maintain it.

Hm... My thought was generating commandRefs today is essentially providing the index number of the batch command (at least for the explicit case, I would need to know more how you plan to implement the implicit case to potentially understand why you are pushing back on this), so I just have been assuming that generating the commandRef is trivial. As for the maintaining part, I am not following your concern. We will allow for backwards compatibility so if you could explain your maintenance concern a little more that would be helpful.

Let me know if a call would be better in discussing this.